### PR TITLE
Gtoc projection

### PR DESCRIPTION
**Pull request by rshanbhageum**
- [X ] Ground Track Oblique Cassini
- [X ] The new projection has been derived by the following paper 
          Stephen Mills, "The ground track oblique Cassini projection used for producing VIIRS mapped imagery," Proc. SPIE 9218,     Earth Observing Systems XIX, 92180C (2 October 2014); [https://doi.org/10.1117/12.2060760](https://doi.org/10.1117/12.2060760)
No change has been made to the documentation for the API in  `docs/source/*.rst` 

Point of contact for GTOC implementation details:  Alessandro Burini (Alessandro.Burini@eumetsat.int)

### CODE REVIEW

File: data/sql/commit.sql [hunk 1]
```diff
@@ -184,6 +184,14 @@ FOR EACH ROW BEGIN
     SELECT RAISE(ABORT, 'missing au_ga_AUSGeoid98.tif')
         WHERE NOT EXISTS(SELECT 1 FROM grid_alternatives WHERE proj_grid_name = 'au_ga_AUSGeoid98.tif');
 
+    -- clause because of buggy entries in EPSG 10.014
```
Comment (rouault) [inline data/sql/commit.sql:187]: this change should be reverted

File: data/sql/customizations.sql [hunk 2]
```diff
@@ -356,4 +356,6 @@ SELECT
     u.scope_code
 FROM grid_transformation gt
 JOIN usage u ON u.object_auth_name = gt.auth_name AND u.object_code = gt.code AND u.object_table_name = 'grid_transformation'
-WHERE method_auth_name = 'EPSG' AND method_code IN ('1088', '1089', '1090', '1091', '1092', '1093', '1094', '1095', '1096', '1097', '1098', '1103') AND gt.deprecated = 0;
+WHERE method_auth_name = 'EPSG' AND method_code IN ('1088', '1089', '1090', '1091', '1092', '1093', '1094', '1095', '1096', '1097', '1098', '1103') AND gt.deprecated = 0
```
Comment (rouault) [inline data/sql/customizations.sql:359]: this change should be reverted

File: data/sql/grid_transformation.sql [hunk 3]
```diff
@@ -984,7 +984,7 @@ INSERT INTO "grid_transformation" VALUES('EPSG','9625','NAD83(MA11) to NAD83(MA1
...
 INSERT INTO "usage" VALUES('EPSG','14463','grid_transformation','EPSG','9626','EPSG','2288','EPSG','1270');
-INSERT INTO "grid_transformation" VALUES('EPSG','9627','NZGD2000 to NZGD2000 + NZVD2009 height (2)','Reversible alternative to NZGD2000 to NZVD2009 height (2) (code 9325).','EPSG','1088','Geog3D to Geog2D+GravityRelatedHeight (gtx)','EPSG','4959','EPSG','9527',0.1,'EPSG','8666','Geoid (height correction) model file','nzgeoid2009.gtx',NULL,NULL,NULL,NULL,'EPSG','4167','LINZ-NZ 2009 gtx',0);
+INSERT INTO "grid_transformation" VALUES('EPSG','9627','NZGD2000 to NZGD2000 + NZVD2009 height (2)','Reversible alternative to NZGD2000 to NZVD2009 height (2) (code 9325).','EPSG','1088','Geog3D to Geog2D+GravityRelatedHeight (gtx)','EPSG','4959','EPSG','4440',0.1,'EPSG','8666','Geoid (height correction) model file','nzgeoid2009.gtx',NULL,NULL,NULL,NULL,'EPSG','4167','LINZ-NZ 2009 gtx',0);
```
Comment (rouault) [inline data/sql/grid_transformation.sql:987]: this change should be reverted

File: data/sql/metadata.sql [hunk 4]
```diff
@@ -9,5 +9,5 @@
 INSERT INTO "metadata" VALUES('DATABASE.LAYOUT.VERSION.MAJOR', 1);
 INSERT INTO "metadata" VALUES('DATABASE.LAYOUT.VERSION.MINOR', 0);
 
-INSERT INTO "metadata" VALUES('EPSG.VERSION', 'v10.015');
-INSERT INTO "metadata" VALUES('EPSG.DATE', '2021-02-11');
+INSERT INTO "metadata" VALUES('EPSG.VERSION', 'v10.014');
```
Comment (rouault) [inline data/sql/metadata.sql:12]: this change should be reverted

File: include/proj/internal/io_internal.hpp [hunk 5]
```diff
@@ -201,10 +201,7 @@ struct projCppContext {
         return auxDbPaths_;
     }
 
-    void setAutoCloseDb(bool autoClose) {
-        autoCloseDb_ = autoClose;
-        autoCloseDbIfNeeded();
-    }
+    void setAutoCloseDb(bool autoClose) { autoCloseDb_ = autoClose; }
```
Comment (rouault) [inline include/proj/internal/io_internal.hpp:204]: this change should be reverted

File: src/Makefile.am [hunk 6]
```diff
@@ -50,7 +50,7 @@ geodtest_LDADD = libproj.la
 
 lib_LTLIBRARIES = libproj.la
 
-libproj_la_LDFLAGS = -no-undefined -version-info 22:0:0
+libproj_la_LDFLAGS = -no-undefined -version-info 21:0:2
```
Comment (rouault) [inline src/Makefile.am:53]: this line should be reverted

File: src/ctx.cpp [hunk 7]
```diff
@@ -87,7 +87,7 @@ void proj_assign_context( PJ* pj, PJ_CONTEXT *ctx )
 pj_ctx pj_ctx::createDefault()
 {
     pj_ctx ctx;
-    ctx.debug_level = PJ_LOG_ERROR;
+    ctx.debug_level = PJ_LOG_NONE;
```
Comment (rouault) [inline src/ctx.cpp:90]: this change should be reverted

File: src/iso19111/metadata.cpp [hunk 8]
```diff
@@ -1124,7 +1124,7 @@ void Identifier::_exportToJSON(JSONFormatter *formatter) const {
 //! @cond Doxygen_Suppress
 static bool isIgnoredChar(char ch) {
     return ch == ' ' || ch == '_' || ch == '-' || ch == '/' || ch == '(' ||
-           ch == ')' || ch == '.' || ch == '&' || ch == ',';
+           ch == ')' || ch == '.' || ch == '&';
```
Comment (rouault) [inline src/iso19111/metadata.cpp:1127]: this change should be reverted

File: src/lib_proj.cmake [hunk 9]
```diff
@@ -309,21 +310,24 @@ set(ALL_LIBPROJ_SOURCES
 )
 set(ALL_LIBPROJ_HEADERS ${HEADERS_LIBPROJ})
 
-# Configuration for the core target "proj"
-proj_target_output_name(proj PROJ_CORE_TARGET_OUTPUT_NAME)
+# Core targets configuration
```
Comment (rouault) [inline src/lib_proj.cmake:313]: all the below changes in that file should be reverted

File: src/proj_internal.h [hunk 10]
```diff
@@ -647,7 +647,7 @@ struct projFileApiCallbackAndData
 struct pj_ctx{
     std::string lastFullErrorMessage{}; // used by proj_context_errno_string
     int     last_errno = 0;
-    int     debug_level = PJ_LOG_ERROR;
+    int     debug_level = 0;
```
Comment (rouault) [inline src/proj_internal.h:650]: this change should be reverted

File: src/projections/gtoc.cpp [hunk 11]
```diff
@@ -0,0 +1,264 @@
+/******************************************************************************
```
Comment (rouault) [inline src/projections/gtoc.cpp:1]: would you mind adding a copyright line ?

File: src/projections/gtoc.cpp [hunk 12]
```diff
@@ -0,0 +1,264 @@
...
+
+PJ *PROJECTION(gtoc) {
+    struct pj_opaque *Q =
+        static_cast<struct pj_opaque *>(calloc(1, sizeof(struct pj_opaque)));
+    if (nullptr == Q)
+        return pj_default_destructor(P, ENOMEM);
+    P->opaque = Q;
+
+    // Earth ellipsoid parameters:
+    double equR = 6378137.0;
```
Comment (rouault) [inline src/projections/gtoc.cpp:213]:
```suggestion
    double equR = P->a;
```

File: src/projections/gtoc.cpp [hunk 13]
```diff
@@ -0,0 +1,264 @@
...
+    double polR = 6356752.3;
```
Comment (rouault) [inline src/projections/gtoc.cpp:214]:
```suggestion
    double polR = P->b;
```

File: src/projections/gtoc.cpp [hunk 14]
```diff
@@ -0,0 +1,264 @@
...
+    // Initializing the trnansformation matrices:
+    // Converting Geotedic to Spheric
+    mat tmp1 = matmult(rotY(Q->theta0), rotX(-Q->az0));
+    mat tmp2 = matmult(rotY(ell2sph(Q->epr2, Q->lat0)), rotZ(-Q->lon0));
+    Q->e2m = matmult(tmp1, tmp2);
+
+    Q->m2e = transpose(Q->e2m);
+
+    // Prevent the use of CLASSIC units by default for outputs:
+    P->right = PJ_IO_UNITS_WHATEVER;
```
Comment (rouault) [inline src/projections/gtoc.cpp:258]: actually is there a good reason to do so ? Reviewing my above suggestions of using P->a / P->b, why not using the unit sphere/ellipsoid with the flattening of the ellipsoid considered ?

File: test/unit/test_c_api.cpp [hunk 15]
```diff
@@ -5185,6 +5185,108 @@ TEST_F(CApi, datum_ensemble) {
 
 // ---------------------------------------------------------------------------
 
+TEST_F(CApi, check_gtoc_projection) {
```
Comment (rouault) [inline test/unit/test_c_api.cpp:5188]: for projections, we rather use the test/gie/ infrastructure. You could add new tests to test/gie/more_builtins.gie for example

**Review by rouault — CHANGES_REQUESTED** [review]

**Review by kbevers — CHANGES_REQUESTED** [review]
Documentation missing. See `/docs/source/operations/projections/merc.rst` for an example.

**Pull request closed without merging**